### PR TITLE
[arkit] Update to Xcode 10 beta 1 (ARKit 2.0)

### PR DESCRIPTION
**Pull request by VincentDondain**
I obsoleted `GetProjectPoint` in favor of `Project` because of the introduction of `Unproject` and based on the API doc https://developer.apple.com/documentation/arkit/arcamera/2923538-projectpoint?language=objc. The `CGPoint` returned is the projection of a point. An other option would have been to name the method `GetProjectedPoint` but I think `Project` is closer to the original and clear enough.

### CODE REVIEW

File: src/arkit.cs [hunk 1]
```diff
@@ -177,9 +210,20 @@ interface ARCamera : NSCopying {
 			get;
 		}
 
+#if !XAMCORE_4_0
+		[Obsolete ("Use 'ProjectPoint' instead.")]
```
Comment (dalexsoto) [inline src/arkit.cs]: Makes sense but I think you meant `Project` instead of `ProjectPoint` 😉

File: src/arkit.cs [hunk 2]
```diff
@@ -1045,6 +1158,128 @@ interface ARDirectionalLightEstimate {
...
+
+		[Export ("exportObjectToURL:previewImage:error:")]
+		bool ExportObject (NSUrl url, [NullAllowed] UIImage previewImage, [NullAllowed] out NSError error);
+
+		[Export ("referenceObjectByApplyingTransform:")]
+		[MarshalDirective (NativePrefix = "xamarin_simd__", Library = "__Internal")]
+		ARReferenceObject GetReferenceObject (Matrix4 transform);
+	}
+
+	[iOS (12,0)]
```
Comment (dalexsoto) [inline src/arkit.cs]: just curious why `[NoWatch, NoTV, NoMac]` isn't present here but it is above??

Comment (dalexsoto) [reply src/arkit.cs]: Same for `ARWorldMap`

File: src/arkit.cs (hunk 1 above)
Comment (spouliot) [inline src/arkit.cs]: should not it be `Project` ?

File: src/arkit.cs [hunk 3]
```diff
@@ -1045,6 +1158,128 @@ interface ARDirectionalLightEstimate {
...
+		[Export ("autoFocusEnabled")]
+		bool AutoFocusEnabled { [Bind ("isAutoFocusEnabled")] get; set; }
+
+		[Export ("planeDetection", ArgumentSemantic.Assign)]
+		ARPlaneDetection PlaneDetection { get; set; }
+	}
+
+	[iOS (12,0)]
+	[NoWatch, NoTV, NoMac]
+	[BaseType (typeof(ARAnchor))]
```
Comment (spouliot) [inline src/arkit.cs]: Considering the `init*` don't accept `nil` is seems unlikely the default constructor should be allowed.

File: src/arkit.cs [hunk 4]
```diff
@@ -1045,6 +1158,128 @@ interface ARDirectionalLightEstimate {
...
+			get;
+		}
+
+		[Export ("rawFeaturePoints", ArgumentSemantic.Strong)]
+		ARPointCloud RawFeaturePoints { get; }
+
+		[Static]
+		[Export ("referenceObjectsInGroupNamed:bundle:")]
+		[return: NullAllowed]
+		NSSet<ARReferenceObject> GetReferenceObjectsInGroup (string name, [NullAllowed] NSBundle bundle);
```
Comment (spouliot) [inline src/arkit.cs]: last part of the selector is the first parameter name
`GetReferenceObjects (string groupName, ...)` seems better

File: src/arkit.cs [hunk 5]
```diff
@@ -1045,6 +1158,128 @@ interface ARDirectionalLightEstimate {
...
+		[Export ("exportObjectToURL:previewImage:error:")]
+		bool ExportObject (NSUrl url, [NullAllowed] UIImage previewImage, [NullAllowed] out NSError error);
+
+		[Export ("referenceObjectByApplyingTransform:")]
+		[MarshalDirective (NativePrefix = "xamarin_simd__", Library = "__Internal")]
+		ARReferenceObject GetReferenceObject (Matrix4 transform);
+	}
+
+	[iOS (12,0)]
+	[BaseType (typeof(ARAnchor))]
```
Comment (spouliot) [inline src/arkit.cs]: the base type has it's default actor marked as not available
unless mentioned explicitly in the headers this should not expose it

**Comment by stephen-hawley** [comment]
:eyes:

**Comment by spouliot** [comment]
xtro files were not updated

File: src/arkit.cs (hunk 1 above)
Comment (VincentDondain) [reply src/arkit.cs]: Ah yes forgot to update that, thanks.

File: src/arkit.cs (hunk 4 above)
Comment (VincentDondain) [reply src/arkit.cs]: Yup read that selector wrong

File: src/arkit.cs (hunk 2 above)
Comment (VincentDondain) [reply src/arkit.cs]: Good catch I missed those.

**Comment by VincentDondain** [comment]
@rolfbjarne I have to revisit the things I added to ARKit.ignore, I'm not quite sure why they are reported yet given the typedef (simd_floatX is typedefed to vector_floatX).

Also what tests were you talking about in your comment? The `Simd` monotouch-tests or ARKit specific tests? I don't recall specific tests for `MarshalDirective`.

**Comment by rolfbjarne** [comment]
@VincentDondain 

> Also what tests were you talking about in your comment? The Simd monotouch-tests or ARKit specific tests? I don't recall specific tests for MarshalDirective.

Just monotouch-tests exercising all API with a MarshalDirective.

**Comment by spouliot** [comment]
Some types might not be user-creatable so doing unit tests might not be possible. @rolfbjarne did you have an idea for them ?

If not let's do tests for the types that are (user createable) inside this PR. For others we better have a sample that exercise the API and, having the PR merged, might also give a chance for more people to try it out.

**Comment by rolfbjarne** [comment]
> Some types might not be user-creatable so doing unit tests might not be possible. 

If they're not user-creatable (and there's no other trivial API that creates them), then we should just skip doing monotouch-tests for them, and instead rely on samples.

**Comment by VincentDondain** [comment]
Comments addressed, I added tests for everything that was user-creatable.

_Bonus: the .arobject 3D model was generated from the real world object bellow._

model3

**Comment by VincentDondain** [comment]
@rolfbjarne @dalexsoto please re-review (:

File: src/arkit.cs [hunk 6]
```diff
@@ -177,9 +210,20 @@ interface ARCamera : NSCopying {
 			get;
 		}
 
+#if !XAMCORE_4_0
+		[Obsolete ("Use 'Project' instead.")]
+		[Wrap ("Project (point, orientation, viewportSize)")]
+		CGPoint GetProjectPoint (Vector3 point, UIInterfaceOrientation orientation, CGSize viewportSize);
```
Comment (rolfbjarne) [inline src/arkit.cs]: This is a breaking change: it's removing the `virtual` specifier (see the api diff for this PR).

**Pull request merged**